# CREDIT CARD APPROVAL PROJECT

This notebook presents a complete machine learning project to predict **credit card approval** using client application data and historical credit behavior.

We'll work with two datasets:

- `application_record.csv`: Contains demographic and financial attributes for each client (one row per client).
- `credit_record.csv`: Contains monthly credit status history per client (multiple rows per client).

The goal is to build a **model** that predicts whether a client should be approved for a credit card, based on their profile and past credit behavior.

### Key steps in this notebook:

1. **Load and inspect the data**
2. **Clean and merge the datasets**
3. **Define a meaningful target variable** (late payment history as a proxy for credit risk for example)
4. **Explore and visualize the data**
5. **Engineer useful features**
6. **Train and evaluate classification models**
7. **Interpret model results and feature importance**

This project will highlight best practices in real-world data science, including dealing with missing values, imbalanced classes, and model evaluation beyond accuracy.


Let's get to it.

In [166]:
# Importing libraries
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [167]:
# Now let's load the datasets
df_application = pd.read_csv("data/application_record.csv")
df_credit = pd.read_csv("data/credit_record.csv")

Once we've done this, we're gonna take a look at both dataframes that we've just imported

In [168]:
df_application

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0       5008804           M            Y               Y             0   
1       5008805           M            Y               Y             0   
2       5008806           M            Y               Y             0   
3       5008808           F            N               Y             0   
4       5008809           F            N               Y             0   
...         ...         ...          ...             ...           ...   
438552  6840104           M            N               Y             0   
438553  6840222           F            N               N             0   
438554  6841878           F            N               N             0   
438555  6842765           F            N               Y             0   
438556  6842885           F            N               Y             0   

        AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0               427500.0               Working               Higher education   
1               427500.0               Working               Higher education   
2               112500.0               Working  Secondary / secondary special   
3               270000.0  Commercial associate  Secondary / secondary special   
4               270000.0  Commercial associate  Secondary / secondary special   
...                  ...                   ...                            ...   
438552          135000.0             Pensioner  Secondary / secondary special   
438553          103500.0               Working  Secondary / secondary special   
438554           54000.0  Commercial associate               Higher education   
438555           72000.0             Pensioner  Secondary / secondary special   
438556          121500.0               Working  Secondary / secondary special   

          NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0             Civil marriage   Rented apartment      -12005          -4542   
1             Civil marriage   Rented apartment      -12005          -4542   
2                    Married  House / apartment      -21474          -1134   
3       Single / not married  House / apartment      -19110          -3051   
4       Single / not married  House / apartment      -19110          -3051   
...                      ...                ...         ...            ...   
438552             Separated  House / apartment      -22717         365243   
438553  Single / not married  House / apartment      -15939          -3007   
438554  Single / not married       With parents       -8169           -372   
438555               Married  House / apartment      -21673         365243   
438556               Married  House / apartment      -18858          -1201   

        FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0                1                1           0           0             NaN   
1                1                1           0           0             NaN   
2                1                0           0           0  Security staff   
3                1                0           1           1     Sales staff   
4                1                0           1           1     Sales staff   
...            ...              ...         ...         ...             ...   
438552           1                0           0           0             NaN   
438553           1                0           0           0        Laborers   
438554           1                1           0           0     Sales staff   
438555           1                0           0           0             NaN   
438556           1                0           1           0     Sales staff   

        CNT_FAM_MEMBERS  
0                   2.0  
1                   2.0  
2                   2.0  
3                   1.0  
4                   1.0  
...                 ...  
438552              1.0  
438553              1.0  
438554              1.0  
43855

As we see, it is a big dataset, with 438557 rows and 18 columns, now let's see if it has any missing values...

In [169]:
df_application.isna().sum()

ID                          0
CODE_GENDER                 0
FLAG_OWN_CAR                0
FLAG_OWN_REALTY             0
CNT_CHILDREN                0
AMT_INCOME_TOTAL            0
NAME_INCOME_TYPE            0
NAME_EDUCATION_TYPE         0
NAME_FAMILY_STATUS          0
NAME_HOUSING_TYPE           0
DAYS_BIRTH                  0
DAYS_EMPLOYED               0
FLAG_MOBIL                  0
FLAG_WORK_PHONE             0
FLAG_PHONE                  0
FLAG_EMAIL                  0
OCCUPATION_TYPE        134203
CNT_FAM_MEMBERS             0
dtype: int64

Great, almost every column is free of missing values, just the `OCCUPATION_TYPE` column has. We're going to take care of it later.

Now let's take a look at the other dataset and see how many missing values it has

In [170]:
df_credit

ID  MONTHS_BALANCE STATUS
0        5001711               0      X
1        5001711              -1      0
2        5001711              -2      0
3        5001711              -3      0
4        5001712               0      C
...          ...             ...    ...
1048570  5150487             -25      C
1048571  5150487             -26      C
1048572  5150487             -27      C
1048573  5150487             -28      C
1048574  5150487             -29      C

[1048575 rows x 3 columns]

This is also a big set, even bigger that the last, with over 1 million rows. Let's look for missing values now...

In [171]:
df_credit.isna().sum()

ID                0
MONTHS_BALANCE    0
STATUS            0
dtype: int64

That is convinient, we have no missing values on this dataset.

Now let's go back and look at the `OCCUPATION_TYPE` column on the first dataset.  
We'll see with what are we dealing with

In [172]:
df_application["OCCUPATION_TYPE"]

0                    NaN
1                    NaN
2         Security staff
3            Sales staff
4            Sales staff
               ...      
438552               NaN
438553          Laborers
438554       Sales staff
438555               NaN
438556       Sales staff
Name: OCCUPATION_TYPE, Length: 438557, dtype: object

In [173]:
df_application["OCCUPATION_TYPE"].value_counts()

OCCUPATION_TYPE
Laborers                 78240
Core staff               43007
Sales staff              41098
Managers                 35487
Drivers                  26090
High skill tech staff    17289
Accountants              15985
Medicine staff           13520
Cooking staff             8076
Security staff            7993
Cleaning staff            5845
Private service staff     3456
Low-skill Laborers        2140
Secretaries               2044
Waiters/barmen staff      1665
Realty agents             1041
HR staff                   774
IT staff                   604
Name: count, dtype: int64

I think the best way to handle this missing values is to fill them with "Unknown" because there's no reason to drop it since it has a lot of values that are not missing

In [174]:
# fill the missing values with "Unknown"
df_application["OCCUPATION_TYPE"] = df_application["OCCUPATION_TYPE"].fillna("Unknown")

In [175]:
# Check if there're still missing values...
df_application.isna().sum()

ID                     0
CODE_GENDER            0
FLAG_OWN_CAR           0
FLAG_OWN_REALTY        0
CNT_CHILDREN           0
AMT_INCOME_TOTAL       0
NAME_INCOME_TYPE       0
NAME_EDUCATION_TYPE    0
NAME_FAMILY_STATUS     0
NAME_HOUSING_TYPE      0
DAYS_BIRTH             0
DAYS_EMPLOYED          0
FLAG_MOBIL             0
FLAG_WORK_PHONE        0
FLAG_PHONE             0
FLAG_EMAIL             0
OCCUPATION_TYPE        0
CNT_FAM_MEMBERS        0
dtype: int64

Now that we managed the missing values, let's look at the datatypes of each column in the dataset...

In [176]:
df_application.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 438557 entries, 0 to 438556
Data columns (total 18 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   ID                   438557 non-null  int64  
 1   CODE_GENDER          438557 non-null  object 
 2   FLAG_OWN_CAR         438557 non-null  object 
 3   FLAG_OWN_REALTY      438557 non-null  object 
 4   CNT_CHILDREN         438557 non-null  int64  
 5   AMT_INCOME_TOTAL     438557 non-null  float64
 6   NAME_INCOME_TYPE     438557 non-null  object 
 7   NAME_EDUCATION_TYPE  438557 non-null  object 
 8   NAME_FAMILY_STATUS   438557 non-null  object 
 9   NAME_HOUSING_TYPE    438557 non-null  object 
 10  DAYS_BIRTH           438557 non-null  int64  
 11  DAYS_EMPLOYED        438557 non-null  int64  
 12  FLAG_MOBIL           438557 non-null  int64  
 13  FLAG_WORK_PHONE      438557 non-null  int64  
 14  FLAG_PHONE           438557 non-null  int64  
 15  FLAG_EMAIL       

In [177]:
df_credit.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 3 columns):
 #   Column          Non-Null Count    Dtype 
---  ------          --------------    ----- 
 0   ID              1048575 non-null  int64 
 1   MONTHS_BALANCE  1048575 non-null  int64 
 2   STATUS          1048575 non-null  object
dtypes: int64(2), object(1)
memory usage: 24.0+ MB


The next steps here are preprocess the data to fit it into a model, so lets do that

First, lets look at the first dataset and see what we can do...

In [178]:
df_application.head(20)

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0   5008804           M            Y               Y             0   
1   5008805           M            Y               Y             0   
2   5008806           M            Y               Y             0   
3   5008808           F            N               Y             0   
4   5008809           F            N               Y             0   
5   5008810           F            N               Y             0   
6   5008811           F            N               Y             0   
7   5008812           F            N               Y             0   
8   5008813           F            N               Y             0   
9   5008814           F            N               Y             0   
10  5008815           M            Y               Y             0   
11  5112956           M            Y               Y             0   
12  6153651           M            Y               Y             0   
13  5008819           M            Y               Y             0   
14  5008820           M            Y               Y             0   
15  5008821           M            Y               Y             0   
16  5008822           M            Y               Y             0   
17  5008823           M            Y               Y             0   
18  5008824           M            Y               Y             0   
19  5008825           F            Y               N             0   

    AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0           427500.0               Working               Higher education   
1           427500.0               Working               Higher education   
2           112500.0               Working  Secondary / secondary special   
3           270000.0  Commercial associate  Secondary / secondary special   
4           270000.0  Commercial associate  Secondary / secondary special   
5           270000.0  Commercial associate  Secondary / secondary special   
6           270000.0  Commercial associate  Secondary / secondary special   
7           283500.0             Pensioner               Higher education   
8           283500.0             Pensioner               Higher education   
9           283500.0             Pensioner               Higher education   
10          270000.0               Working               Higher education   
11          270000.0               Working               Higher education   
12          270000.0               Working               Higher education   
13          135000.0  Commercial associate  Secondary / secondary special   
14          135000.0  Commercial associate  Secondary / secondary special   
15          135000.0  Commercial associate  Secondary / secondary special   
16          135000.0  Commercial associate  Secondary / secondary special   
17          135000.0  Commercial associate  Secondary / secondary special   
18          135000.0  Commercial associate  Secondary / secondary special   
19          130500.0               Working              Incomplete higher   

      NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0         Civil marriage   Rented apartment      -12005          -4542   
1         Civil marriage   Rented apartment      -12005          -4542   
2                Married  House / apartment      -21474          -1134   
3   Single / not married  House / apartment      -19110          -3051   
4   Single / not married  House / apartment      -19110          -3051   
5   Single / not married  House / apartment      -19110          -3051   
6   Single / not married  House / apartment      -19110          -3051   
7              Separated  House / apartment      -22464         365243   
8              Separated  House / apartment      -22464         365243   
9              Separated  House / apartment      -22464         365243   
10               Married  House / apartment      -16872           -769   
11               Married  House 

Here we see that for example, the columns `"DAYS_BIRTH"` and `"DAYS_EMPLOYED"` are negative values, which it's not convenient, we should change it to be positive and improving readability by dividing it by 365. This will be stored in a new column and we'll drop the original ones.  
`"DAYS_BIRTH"`, `"DAYS_EMPLOYED"`: These represent days before the current day
But in `DAYS_EMPLOYED` there's a flag that indicates us that that person is currently unemployed, that's the value **365243**, so we have to handle that before anything.

### Getting data ready

In [179]:
# First, let's transform the DAYS_BIRTH and DAYS_EMPLOYED 
df_application["AGE"] = np.abs(df_application["DAYS_BIRTH"])//365
df_application["SENIORITY"] = np.abs(df_application["DAYS_EMPLOYED"])//365

df_application = df_application.drop(['DAYS_BIRTH', 'DAYS_EMPLOYED'], axis=1)

# Let's handle the FLAG now (now it'll be 1000 on SENIORITY)
df_application["SENIORITY"] = df_application["SENIORITY"].replace(1000, np.nan)
# Now we can impute the mean of the valid values
df_application["NO_EMPLOYMENT_FLAG"] = df_application["SENIORITY"].isna().astype(int)
df_application["SENIORITY"] = df_application["SENIORITY"].fillna(np.mean(df_application["SENIORITY"]))

df_application

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0       5008804           M            Y               Y             0   
1       5008805           M            Y               Y             0   
2       5008806           M            Y               Y             0   
3       5008808           F            N               Y             0   
4       5008809           F            N               Y             0   
...         ...         ...          ...             ...           ...   
438552  6840104           M            N               Y             0   
438553  6840222           F            N               N             0   
438554  6841878           F            N               N             0   
438555  6842765           F            N               Y             0   
438556  6842885           F            N               Y             0   

        AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0               427500.0               Working               Higher education   
1               427500.0               Working               Higher education   
2               112500.0               Working  Secondary / secondary special   
3               270000.0  Commercial associate  Secondary / secondary special   
4               270000.0  Commercial associate  Secondary / secondary special   
...                  ...                   ...                            ...   
438552          135000.0             Pensioner  Secondary / secondary special   
438553          103500.0               Working  Secondary / secondary special   
438554           54000.0  Commercial associate               Higher education   
438555           72000.0             Pensioner  Secondary / secondary special   
438556          121500.0               Working  Secondary / secondary special   

          NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  FLAG_MOBIL  FLAG_WORK_PHONE  \
0             Civil marriage   Rented apartment           1                1   
1             Civil marriage   Rented apartment           1                1   
2                    Married  House / apartment           1                0   
3       Single / not married  House / apartment           1                0   
4       Single / not married  House / apartment           1                0   
...                      ...                ...         ...              ...   
438552             Separated  House / apartment           1                0   
438553  Single / not married  House / apartment           1                0   
438554  Single / not married       With parents           1                1   
438555               Married  House / apartment           1                0   
438556               Married  House / apartment           1                0   

        FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  AGE  \
0                0           0         Unknown              2.0   32   
1                0           0         Unknown              2.0   32   
2                0           0  Security staff              2.0   58   
3                1           1     Sales staff              1.0   52   
4                1           1     Sales staff              1.0   52   
...            ...         ...             ...              ...  ...   
438552           0           0         Unknown              1.0   62   
438553           0           0        Laborers              1.0   43   
438554           0           0     Sales staff              1.0   22   
438555           0           0         Unknown              2.0   59   
438556           1           0     Sales staff              2.0   51   

        SENIORITY  NO_EMPLOYMENT_FLAG  
0       12.000000                   0  
1       12.000000                   0  
2        3.000000                   0  
3        8.000000                   0  
4        8.000000                   0  
...           ...                 ...  
438552   6.684196                   1  
43855

In [180]:
# Now we have to map the CODE_GENDER, FLAG_OWN_CAR and FLAG_OWN_REALTY columns
df_application['CODE_GENDER'] = df_application['CODE_GENDER'].map({'M': 0, 'F': 1})
df_application['FLAG_OWN_CAR'] = df_application['FLAG_OWN_CAR'].map({'N': 0, 'Y': 1})
df_application['FLAG_OWN_REALTY'] = df_application['FLAG_OWN_REALTY'].map({'N': 0, 'Y': 1})
df_application

ID  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  CNT_CHILDREN  \
0       5008804            0             1                1             0   
1       5008805            0             1                1             0   
2       5008806            0             1                1             0   
3       5008808            1             0                1             0   
4       5008809            1             0                1             0   
...         ...          ...           ...              ...           ...   
438552  6840104            0             0                1             0   
438553  6840222            1             0                0             0   
438554  6841878            1             0                0             0   
438555  6842765            1             0                1             0   
438556  6842885            1             0                1             0   

        AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0               427500.0               Working               Higher education   
1               427500.0               Working               Higher education   
2               112500.0               Working  Secondary / secondary special   
3               270000.0  Commercial associate  Secondary / secondary special   
4               270000.0  Commercial associate  Secondary / secondary special   
...                  ...                   ...                            ...   
438552          135000.0             Pensioner  Secondary / secondary special   
438553          103500.0               Working  Secondary / secondary special   
438554           54000.0  Commercial associate               Higher education   
438555           72000.0             Pensioner  Secondary / secondary special   
438556          121500.0               Working  Secondary / secondary special   

          NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  FLAG_MOBIL  FLAG_WORK_PHONE  \
0             Civil marriage   Rented apartment           1                1   
1             Civil marriage   Rented apartment           1                1   
2                    Married  House / apartment           1                0   
3       Single / not married  House / apartment           1                0   
4       Single / not married  House / apartment           1                0   
...                      ...                ...         ...              ...   
438552             Separated  House / apartment           1                0   
438553  Single / not married  House / apartment           1                0   
438554  Single / not married       With parents           1                1   
438555               Married  House / apartment           1                0   
438556               Married  House / apartment           1                0   

        FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  AGE  \
0                0           0         Unknown              2.0   32   
1                0           0         Unknown              2.0   32   
2                0           0  Security staff              2.0   58   
3                1           1     Sales staff              1.0   52   
4                1           1     Sales staff              1.0   52   
...            ...         ...             ...              ...  ...   
438552           0           0         Unknown              1.0   62   
438553           0           0        Laborers              1.0   43   
438554           0           0     Sales staff              1.0   22   
438555           0           0         Unknown              2.0   59   
438556           1           0     Sales staff              2.0   51   

        SENIORITY  NO_EMPLOYMENT_FLAG  
0       12.000000                   0  
1       12.000000                   0  
2        3.000000                   0  
3        8.000000                   0  
4        8.000000                   0  
...           ...                 ...  
438552   

Then we must merge the dataframes so all the data is on just one DF

In [181]:
# Create a binary target: 1 if ever had STATUS in ['2','3','4','5'], else 0
# Filter and aggregate credit record by ID
df_credit['bad_credit'] = df_credit['STATUS'].isin(['2', '3', '4', '5']).astype(int)
credit_agg = df_credit.groupby('ID')['bad_credit'].max().reset_index()

# Merge with application_df on ID
df = pd.merge(df_application, credit_agg, on='ID')
df.head()

ID  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  CNT_CHILDREN  \
0  5008804            0             1                1             0   
1  5008805            0             1                1             0   
2  5008806            0             1                1             0   
3  5008808            1             0                1             0   
4  5008809            1             0                1             0   

   AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0          427500.0               Working               Higher education   
1          427500.0               Working               Higher education   
2          112500.0               Working  Secondary / secondary special   
3          270000.0  Commercial associate  Secondary / secondary special   
4          270000.0  Commercial associate  Secondary / secondary special   

     NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  FLAG_MOBIL  FLAG_WORK_PHONE  \
0        Civil marriage   Rented apartment           1                1   
1        Civil marriage   Rented apartment           1                1   
2               Married  House / apartment           1                0   
3  Single / not married  House / apartment           1                0   
4  Single / not married  House / apartment           1                0   

   FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  AGE  SENIORITY  \
0           0           0         Unknown              2.0   32       12.0   
1           0           0         Unknown              2.0   32       12.0   
2           0           0  Security staff              2.0   58        3.0   
3           1           1     Sales staff              1.0   52        8.0   
4           1           1     Sales staff              1.0   52        8.0   

   NO_EMPLOYMENT_FLAG  bad_credit  
0                   0           0  
1                   0           0  
2                   0           0  
3                   0           0  
4                   0           0

Now that we have everything almost settled, we have to preprocess the data through a sklear pipeline to ensure that there won't be any data leakages, and we're going to do it this way...

In [ ]:
# Importing 
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder

# Setting up random seed
np.random.seed(42)

# We have to divide the data into categorical features and nuemerical features

## Model Building

Okay we are in a classification problem, but how do we know which model to use?  
Since this question it's imposible to answer without tests, let's test some models ans see how they perform...

In [148]:
# importing the different models
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier

In [149]:
# setting up a random seed
np.random.seed(42)

# Importing and splitting the data into train and test
from sklearn.model_selection import train_test_split

# Split into X & Y
X = df.drop("BAD_CREDIT", axis=1)
y = df["BAD_CREDIT"]

# Split into train and test 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

KeyError: "['BAD_CREDIT'] not found in axis"

In [ ]:
# Now lets make a dictionary with each model and see which has the best score
# to score these models we're going to use the auc-roc curve and a cross-val-score
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import cross_val_score

models ={
    "Logistic": LogisticRegression(),
    "RandomForest" : RandomForestClassifier(),
    "KNN": KNeighborsClassifier()
}

for name, model in models.items():
    model.fit(X_train,y_train)
    y_proba = model.predict_proba(X_test)[:,1] # for AUC
    auc = roc_auc_score(y_test, y_proba)
    print(f"{name}: ROC AUC = {auc:.4f}")



ValueError: could not convert string to float: 'Working'

Now that we know that the `RandomForestClassifier` it's the best performer, let's work on it

In [ ]:
# build the model
clf = RandomForestClassifier()

# Let's see the model score with the cross_val_score
cross_val_score(clf, X, y, cv = 5)

array([0.06623697, 0.94569391, 0.95254423, 0.93828007, 0.92319298])